File: src/main/java/processes/ProcessHandler.java
package processes;

import java.util.*;
import java.util.stream.Collectors;

public class ProcessHandler {
    public static final List<String> DEFAULT_HIDDEN_PROCESSES =
            List.of("System32", "Nvidia", "SystemApps", "wallpaper",
                    "Razer", "Native Instruments", "xboxGam", "Microsoft.ZuneVideo", "Settings", "GWSL",
                    "Keyboard Chattering Fix", "YourPhone", "webhelper", "Driver", "Gaomon", "Git", "fsnotifier");
    public static List<String> hiddenProcesses = new ArrayList<>(DEFAULT_HIDDEN_PROCESSES);
    public static List<Process> reducedProcessList = null;

    public static Set<String> blacklisted = new HashSet<>();

    //TODO this might not work with lastAllProcess.equals(), because the ordering of ProcessHandle.allProcesses() is
    // unclear
    private static List<ProcessHandle> lastAllProcesses = ProcessHandle.allProcesses().collect(Collectors.toList());

    public static List<Process> computeReducedProcessList(String user, List<String> cmdBlacklist, boolean force) {
        final List<ProcessHandle> newerLastAllProcesses = ProcessHandle.allProcesses().collect(Collectors.toList());
        if (!force && reducedProcessList != null && lastAllProcesses.equals(newerLastAllProcesses)) {
            return reducedProcessList;
        }
        try {
            lastAllProcesses = newerLastAllProcesses;
            reducedProcessList = lastAllProcesses.stream()
                    .map(Process::new)
                    .distinct()
                    .filter(proc -> proc.user().toLowerCase().contains(user.toLowerCase()))
                    .filter(proc -> cmdBlacklist.stream().map(String::toLowerCase)
                            .noneMatch(item -> proc.command().toLowerCase().contains(item)))
                    .collect(Collectors.toList());
        } catch (Exception e) {
            //FIXME: Sometimes:tm:, a concurrent modification Exception occurs here, if the cmdBlacklist is being
            // modified while its stream is being read. This is probably fixable by either locking it with a
            // synchronized block, using a blocking collection, or making a local shallow copy, the elements of which
            // won't be concurrently modified by the user.
            // The workaround, which is much easier, is just returning an empty list for a second, which will
            // be replaced <1 second afterwards
            return Collections.emptyList();
        }
        return reducedProcessList;
    }

    public static List<Process> computeFilteredProcessList(boolean force) {
        return ProcessHandler.computeReducedProcessList(force).stream()
                .filter(proc -> !ProcessHandler.blacklisted.contains(proc.command())).collect(Collectors.toList());
    }

    public static List<Process> computeReducedProcessList(String user) {
        return computeReducedProcessList(user, hiddenProcesses, false);
    }

    public static List<Process> computeReducedProcessList() {
        return computeReducedProcessList(System.getProperty("user.name"), hiddenProcesses, false);
    }

    public static List<Process> computeReducedProcessList(boolean force) {
        return computeReducedProcessList(System.getProperty("user.name"), hiddenProcesses, force);
    }

    public static List<Process> computeReducedProcessList(List<String> cmdBlacklist) {
        return computeReducedProcessList(System.getProperty("user.name"), cmdBlacklist, false);
    }

    public static List<Process> getDisallowedProcessesThatAreRunning() {
        return computeReducedProcessList(true).stream()
                .filter(process -> blacklisted.contains(process.command()))
                .collect(Collectors.toList());
    }

    public static Collection<? extends Process> blacklistedProcesses() {
        if (blacklisted == null) {
            return Collections.emptySet();
        }
        return blacklisted.stream().map(Process::new).collect(Collectors.toList());
    }

    public static void resetHiddenProcesses() {
        hiddenProcesses = new ArrayList<>(DEFAULT_HIDDEN_PROCESSES);
    }
}

File: src/main/kotlin/processes/ProcessHandler.kt
package processes

import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.snapshots.SnapshotStateList
import androidx.compose.runtime.toMutableStateList
import processes.implementations.ProcessIdentifier
import processes.implementations.RunningProcess
import java.util.stream.Stream
import kotlin.streams.asSequence
import kotlin.streams.toList

object ProcessHandler {
    val DEFAULT_HIDDEN_PROCESSES: Set<String> = setOf(
        "System32", "Nvidia", "SystemApps", "wallpaper",
        "Razer", "Native Instruments", "xboxGam", "Microsoft.ZuneVideo", "Settings", "GWSL",
        "Keyboard Chattering Fix", "YourPhone", "webhelper", "Driver", "Gaomon", "Git", "fsnotifier",
        "manager", "launcher", "daemon", "system", "proxy", "kde","qt"
    )
    var blacklisted: SnapshotStateList<Process> = mutableStateListOf()
    var hiddenProcesses: MutableList<Process> = resetHiddenProcesses()

    val afterKill: (RunningProcess) -> Unit = { proc: RunningProcess ->
        allProcesses.remove(proc.command)
    }

    //:blobdoubt:
    private var allProcesses: MutableMap<String, RunningProcess> =
        ProcessHandle.allProcesses().asSequence().associate { handle ->
            (handle.info().command().orElse("")) to (RunningProcess(handle, afterKill))
        }.toMutableMap()

    @JvmOverloads
    fun computeReducedProcessList(
        user: String = System.getProperty("user.name"),
        cmdBlacklist: List<Process> = hiddenProcesses,
    ): Stream<RunningProcess>? {
        val runningProcesses =
            ProcessHandle.allProcesses().asSequence()
                .associateBy { processHandle -> processHandle.info().command().orElse("") }

        synchronized(allProcesses) {
            runningProcesses.values.parallelStream().forEach { handle ->
                val handleCommand = handle.info().command().orElse("")
                allProcesses.computeIfAbsent(handleCommand) {
                    RunningProcess(handle, afterKill)
                }
            }
        }
        val user = user.toLowerCase()
        synchronized(cmdBlacklist) {
            return allProcesses.values
                .parallelStream()
                .distinct()
                .filter { proc ->
                    proc.user.contains(user)
                }
                .filter { proc ->
                    cmdBlacklist.none { otherProcess ->
                        proc == otherProcess || proc.command.contains(
                            otherProcess.stringRepresentation,
                            ignoreCase = true
                        )
                    }
                }
                .filter {
                    runningProcesses.keys.contains(it.command)
                }
        }
    }

    fun computeFilteredProcessList(stream: Stream<RunningProcess>? = computeReducedProcessList()): List<RunningProcess> {
        return stream?.filter { proc: Process ->
            !blacklisted.contains(proc)
        }?.toList() ?: emptyList()
    }

    val disallowedProcessesThatAreRunning: List<RunningProcess>
        get() {
            synchronized(blacklisted) {
                return computeReducedProcessList()
                    ?.filter { process: Process ->
                        blacklisted.contains(process)
                    }?.toList() ?: emptyList()
            }
        }

    fun resetHiddenProcesses(): MutableList<Process> {
        hiddenProcesses = DEFAULT_HIDDEN_PROCESSES
            .map { name -> ProcessIdentifier(stringRepresentation = name) }
            .toMutableStateList()
        return hiddenProcesses.toMutableStateList()
    }
}
